### PR TITLE
tesseract-lang 4.0.0 (new formula)

### PR DESCRIPTION
**Pull request by albertosottile**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?
-----

Following the indications in PR #36786, this PR includes a new `tesseract-lang` formula that depends on `tesseract` and installs all the language data files (besides English and osd, needed to run the software) in its `share` folder. 

The original `tesseract` formula is also updated to embed only the essential data files, and amended to search for language data files in `#{HOMEBREW_PREFIX}/share`. Symlinks do the rest of the trick.

In this way, the installed size of `tesseract` drops to 30 MB, while the size of `tesseract-lang`is about 650 MB. 

The `test` block of both formulas has also been updated to perform an actual OCR task on an example taken from the upstream test repository. The test is made using only English in `tesseract`and using also German in `tesseract-lang`.

### CODE REVIEW

**Comment by albertosottile** [comment]
Side note: I used `inreplace` in the updated `tesseract` formula to show that only the makefiles relative to `tessdata` are affected by this change, but of course the three replaces could be removed in favor of passing ` datarootdir=#{share}` to `make install`.

File: Formula/tesseract-lang.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  end
+
+  test do
+    resource "tests" do
+      url "https://github.com/tesseract-ocr/test.git"
+    end
+
+    resource("tests").stage do
+      system "#{Formula["tesseract"].bin}/tesseract", "./testing/eurotext.tif", "./output", "-l", "eng+deu"
+      assert_equal "über den faulen Hund. Le renard brun\n", shell_output("sed -n 6p ./output.txt")
```
Comment (retokromer) [inline Formula/tesseract-lang.rb]: I would suggest to put also for the French part "something" that uses more that "English" letters, e.g. `Académie française`.

Comment (albertosottile) [reply Formula/tesseract-lang.rb]: Do you mean using another image from the test repository? Any suggestions?

Comment (retokromer) [reply Formula/tesseract-lang.rb]: Yes, something that uses more specific French letters than `Le renard brun`. (In the German part you have at least one Umlaut: `über`.)

Comment (albertosottile) [reply Formula/tesseract-lang.rb]: I could add another line of the same `eurotext.tif` image, but that Umlaut is exactly the point of this test. The fact that `tesseract` correctly used the `deu` language to recognize the ū proves that the language packs were installed correctly. I do not think that is advantageous to test also for French (or for one of the other 160 languages installed), but I can add more languages to this block if Homebrew maintainers think that is necessary. After all, none of these tests existed in the current version of the formula.

File: Formula/tesseract-lang.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  depends_on "tesseract"
+
+  def install
+    rm "eng.traineddata"
+    rm "osd.traineddata"
+    (share/"tessdata").install Dir.glob("./*")
+  end
+
+  test do
+    resource "tests" do
```
Comment (fxcoudert) [inline Formula/tesseract-lang.rb]: - The resource should be defined at the top-level, not inside `test` block.
- Let's not include the whole repo as resource, but only the file you really need (`eurotext.tif`).

File: Formula/tesseract-lang.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  homepage "https://github.com/tesseract-ocr/tessdata_fast/"
+  url "https://github.com/tesseract-ocr/tessdata_fast/archive/4.0.0.tar.gz"
+  sha256 "f1b71e97f27bafffb6a730ee66fd9dc021afc38f318fdc80a464a84a519227fe"
+
+  depends_on "tesseract"
+
+  def install
+    rm "eng.traineddata"
+    rm "osd.traineddata"
+    (share/"tessdata").install Dir.glob("./*")
```
Comment (fxcoudert) [inline Formula/tesseract-lang.rb]:
```suggestion
    (share/"tessdata").install Dir["*"]
```

File: Formula/tesseract-lang.rb (hunk 1 above)
Comment (fxcoudert) [reply Formula/tesseract-lang.rb]: The test looks fine as it is.
Please use `assert_match` directly on the content of output, rather than call `sed`.

File: Formula/tesseract.rb [hunk 4]
```diff
@@ -49,15 +44,29 @@ def install
     ENV.cxx11
 
     system "./autogen.sh"
-    system "./configure", "--prefix=#{prefix}", "--disable-dependency-tracking"
+    system "./configure", "--prefix=#{prefix}", "--disable-dependency-tracking", "--datarootdir=#{HOMEBREW_PREFIX}/share"
+
+    system "make"
+
+    inreplace "tessdata/Makefile", "datarootdir = #{HOMEBREW_PREFIX}/share", "datarootdir = #{share}"
```
Comment (fxcoudert) [inline Formula/tesseract.rb]: Why is that necessary?

File: Formula/tesseract.rb [hunk 5]
```diff
@@ -49,15 +44,29 @@ def install
...
 
     resource("snum").stage { mv "snum.traineddata", share/"tessdata" }
-    resource("tessdata").stage { mv Dir["*"], share/"tessdata" }
+    resource("eng").stage { mv "eng.traineddata", share/"tessdata" }
+    resource("osd").stage { mv "osd.traineddata", share/"tessdata" }
   end
 
   test do
-    assert_match version.to_s, shell_output("#{bin}/tesseract -v 2>&1")
+    resource "tests" do
```
Comment (fxcoudert) [inline Formula/tesseract.rb]: - The resource should be defined at the top-level, not inside `test` block.
- Let's not include the whole repo as resource, but only the file you really need (`eurotext.tif`).

**Review by fxcoudert — CHANGES_REQUESTED** [review]
Looking good. Please bump the revision number of the main formula, and add a `caveats` so users are made aware that extra languages are available under another formula.

File: Formula/tesseract-lang.rb (hunk 1 above)
Comment (retokromer) [reply Formula/tesseract-lang.rb]: I understand, thank you.

Comment (albertosottile) [reply Formula/tesseract-lang.rb]: @fxcoudert I am not sure I understood your suggestion. I removed `sed` in favor of `cat` and used `assert_match`, is that ok?

File: Formula/tesseract.rb (hunk 4 above)
Comment (albertosottile) [reply Formula/tesseract.rb]: As I commented above https://github.com/Homebrew/homebrew-core/pull/36988#issuecomment-463747370, it was not absolutely necessary to use the `inreplace` functions. The `datarootdir` must be corrected before `make install` to avoid a permission error, but the same effect could be obtained by passing  `datarootdir=#{share}` to `make install`. I used the `inreplace` functions to explicitly show which makefiles were affected by this change. If you prefer, I could remove them in favor of passing the argument as discussed above.

File: Formula/tesseract-lang.rb [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  def install
+    rm "eng.traineddata"
+    rm "osd.traineddata"
+    (share/"tessdata").install Dir["*"]
+  end
+
+  test do
+    resource("testfile").stage do
+      system "#{Formula["tesseract"].bin}/tesseract", "./eurotext.tif", "./output", "-l", "eng+deu"
+      assert_match "über den faulen Hund. Le renard brun\n", shell_output("cat ./output.txt")
```
Comment (fxcoudert) [inline Formula/tesseract-lang.rb]:
```suggestion
      assert_match "über den faulen Hund. Le renard brun\n", File.read("output.txt")
```

File: Formula/tesseract.rb [hunk 7]
```diff
@@ -49,15 +50,31 @@ def install
...
+    This formula containes only the "eng", "osd", and "snum" language data files.
+    If you need all the other supported languages, `brew install tesseract-lang`.
+  EOS
   end
 
   test do
-    assert_match version.to_s, shell_output("#{bin}/tesseract -v 2>&1")
+    resource("testfile").stage do
+      system "#{Formula["tesseract"].bin}/tesseract", "./eurotext.tif", "./output", "-l", "eng"
+      assert_match "The (quick) [brown] {fox} jumps!\n", shell_output("cat ./output.txt")
```
Comment (fxcoudert) [inline Formula/tesseract.rb]: Same here

**Review by fxcoudert — CHANGES_REQUESTED** [review]

File: Formula/tesseract.rb (hunk 4 above)
Comment (fxcoudert) [reply Formula/tesseract.rb]: Yes, please pass the argument instead, I think it would be more robust in the future (if another makefile starts using `datarootdir`). Also please add a short comment as to why we're doing this.

**Comment by fxcoudert** [comment]
Finally (and I believe it will be ready to merge): please squash everything into two commits, one per formula (`tesseract-lang 4.0.0 (new formula)` and `tesseract: remove languages`).

**Comment by albertosottile** [comment]
@fxcoudert Everything should be ok now, thanks for your review. Anything else?

**Comment by albertosottile** [comment]
Done. Anything else that should be done?

**Review by fxcoudert — APPROVED** [review]

**Comment by fxcoudert** [comment]
Many thanks @albertosottile for taking care of this!

**Comment by fxcoudert** [comment]
For reference: the compressed `tesseract` bottles have gone back from 339.9 MB to 11.8 MB, while the new `tesseract-lang` bottles are 328.1 MB.
